### PR TITLE
Add upgrade guide to the documentation

### PR DESCRIPTION
**Pull request by dean0x7d**
The basic idea here is to have a separate documentation page which sits right below the changelog, but goes into more detail about changes which specifically affect version upgrades. It can spell out what to watch out for and what the optimal upgrade path is. This is only a small number of items per version, but ones that would feel weird to describe in detail in the changelog. This would hopefully avoid some confusion going forward (e.g. #909).

The sections about hidden visibility and local STL containers feel a bit verbose. Let me know your thoughts about those and the whole thing.

### CODE REVIEW

File: docs/upgrade.rst [hunk 1]
```diff
@@ -0,0 +1,317 @@
...
+localize all common type bindings in order to avoid conflicts with
+third-party modules.
+
+
+New syntax for custom constructors
+----------------------------------
+
+The old placement-new custom constructors are still valid, but the new approach
+greatly improves type safety (placement-new can be called accidentally with an
+incompatible type, without any compiler errors or warnings). See
```
Comment (jagerman) [inline docs/upgrade.rst]: ... with an incompatible type, or can be called multiple times on the same object if not careful with the Python-side `__init__` calls.

File: docs/upgrade.rst [hunk 2]
```diff
@@ -0,0 +1,317 @@
...
+
+    // old
+    py::class<Foo>(m, "Foo")
+        .def("__init__", [](Foo &self, ...) {
+            new (&self) Foo(...); // uses placement-new
+        });
+
+    // new
+    py::class<Foo>(m, "Foo")
+        .def(py::init([](...) {
```
Comment (jagerman) [inline docs/upgrade.rst]: Suggest: ``// Note: no `self` argument`` to draw attention to that distinction.

File: docs/upgrade.rst [hunk 3]
```diff
@@ -0,0 +1,317 @@
...
+
+Deprecation of some ``py::object`` APIs
+---------------------------------------
+
+The equality comparison operator for ``py::object`` only compares the
+underlying pointers, therefore ``obj1 == obj2`` in C++ is equivalent to the
+expression ``obj1 is obj2`` in Python. To avoid confusion, the equality
+operator now emits a deprecation warning (in the future, it can be replaced
+with proper rich object comparison). To compare objects by pointer, use
+``obj1.is(obj2)``.
```
Comment (jagerman) [inline docs/upgrade.rst]: I think this should be rewritten to mention `is` first (otherwise it seems to be suggesting that `==` is the right way to do it if I don't read the paragraph in full).

```
To compare ``py::object``s by pointer, you should now use ``obj1.is(obj2)``,
which is equivalent to ``obj1 is obj2`` in Python.  Previously pybind11 used
the `==` operator for this (`obj1 == obj2`), but that could be confusing and
is now deprecated (so that it can eventually be replaced with proper rich
object comparison in a future release).
```

File: docs/upgrade.rst [hunk 4]
```diff
@@ -0,0 +1,317 @@
...
+v2.0
+====
+
+Breaking changes in ``py::class_``
+----------------------------------
+
+These changes were necessary to make type definitions in pybind11
+future-proof, to support PyPy via its ``cpyext`` mechanism (`#527
+<https://github.com/pybind/pybind11/pull/527>`_), and to improve efficiency
+(`rev. 86d825 <https://github.com/pybind/pybind11/commit/86d825>`_).
```
Comment (jagerman) [inline docs/upgrade.rst]: The referenced change (86d825 - moving trampolines declaration into the `class_<...>` template types) looks like it was already in 1.8.  Should that be a different commit?

Comment (jagerman) [inline docs/upgrade.rst]: Oops, never mind, I was misreading the commit status.

File: docs/upgrade.rst [hunk 5]
```diff
@@ -0,0 +1,317 @@
...
+           .def(py::init<...>())
+           .def_buffer(...);
+
+2. Classes which include static properties (e.g. ``def_readwrite_static()``)
+   must now include the ``py::metaclass()`` attribute. Note: this requirement
+   has since been removed in v2.1. If you're upgrading from 1.x, it's
+   recommended to skip directly to v2.1 or newer.
+
+3. This version of pybind11 uses a redesigned mechnism for instantiating
+   trempoline classes that are used to override virtual methods from within
```
Comment (jagerman) [inline docs/upgrade.rst]: trempoline -> trampoline

File: docs/upgrade.rst [hunk 6]
```diff
@@ -0,0 +1,317 @@
...
+       py::class_<TrampolineClass>("MyClass")
+           .alias<MyClass>()
+           ...
+
+       // new v2.x syntax
+       py::class_<MyClass, TrampolineClass>("MyClass")
+           ...
+
+   Importantly, both the original and the trampoline class are now specified
+   as an arguments (in arbitrary order) to the ``py::class_`` template, and
```
Comment (jagerman) [inline docs/upgrade.rst]: The "in arbitrary order" here is a bit confusing as it seems to be applying to "the original and the trampoline", but of course the original class still has to be first.  Perhaps: "(mixed in arbitrary order with other arguments such as a base class or holder type)".

**Comment by jagerman** [comment]
I left a few minor wording comments, but overall I think this is a great idea.

File: docs/upgrade.rst (hunk 3 above)
Comment (dean0x7d) [inline docs/upgrade.rst]: Good point. The original was awkwardly worded.

File: docs/upgrade.rst [hunk 7]
```diff
@@ -0,0 +1,322 @@
...
+The namespace-scope hidden visibility is done automatically in pybind11's
+headers and it's generally transparent to users. It ensures that:
+
+* Modules compiled with different pybind11 versions don't clash with each other.
+
+* Some new features, like ``py::module_local`` bindings, can work as intended.
+
+The ``-fvisibility=hidden`` flag applies the same visibility to user bindings
+outside of the ``pybind11`` namespace. It's now set automatic by pybind11's
+CMake and Python build systems, but this needs to be done manually by users
```
Comment (wjakob) [inline docs/upgrade.rst]: I'm confused by this paragraph -- didn't the CMake build system always set -fvisibility=hidden on OSX & Linux?

Comment (wjakob) [inline docs/upgrade.rst]: Aha, that point clarified itself a few lines below :)

File: docs/upgrade.rst [hunk 8]
```diff
@@ -0,0 +1,322 @@
...
+* GCC >= 4.8
+* clang >= 3.3 (appleclang >= 5.0)
+* MSVC >= 2015u3
+* Intel C++ >= 15.0
+
+
+Deprecation of the ``py::metaclass()`` attribute
+------------------------------------------------
+
+It's no longer required for binding classes with static properties.
```
Comment (wjakob) [inline docs/upgrade.rst]: Hmm, I think this is too strong. We didn't deprecate the ``py::metaclass(handle)`` constructors, which permits creating pybind11-bound classes with custom metaclasses (fairly powerful IMO).

How about the following instead? The ``py::metaclass`` attribute is no longer required for binding classes with static properties. It is still needed in rare cases, e.g.  to set a custom metaclasses for pybind11-bound data structures.

**Comment by wjakob** [comment]
This is a great idea -- very useful!

File: docs/upgrade.rst (hunk 8 above)
Comment (dean0x7d) [inline docs/upgrade.rst]: It did read a bit like `py::metaclass` was completely deprecated instead of just the zero-parameter version. I've rewritten it now.
